pkg/registry/core/service/storage/rest.go

```diff
@@ -521,6 +536,35 @@ func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableO
 	return r.services.ConvertToTable(ctx, object, tableOptions)
 }
 
+// When allocating we always use BySpec, when releasing we always use ByClusterIP
+func (r *REST) getAllocatorByClusterIP(service *api.Service) ipallocator.Interface {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) || r.secondaryServiceIPs == nil {
+		return r.serviceIPs
+	}
+
+	secondaryAllocatorCIDR := r.secondaryServiceIPs.CIDR()
+	if netutil.IsIPv6String(service.Spec.ClusterIP) && netutil.IsIPv6CIDR(&secondaryAllocatorCIDR) {
+		return r.secondaryServiceIPs
+	}
+
+	return r.serviceIPs
+}
+
+func (r *REST) getAllocatorBySpec(service *api.Service) ipallocator.Interface {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) ||
+		service.Spec.ServiceIPFamily == api.ServiceIPFamilyClusterDefault ||
+		r.secondaryServiceIPs == nil {
```

Comment:
This will disregard the user's request.  I think this should be an error - user asked for v6 in a v4-only cluster.  Or will that get caught elsewhere?

Reply:
It is caught at the allocation side. the allocator will not allocate an IP and will broadcast an error event on the service.